### ISSUE TITLE
Consider removing `--manager` from `swarm join`

### ISSUE THREAD

**Issue by aluzzardi** [docker/docker#24084]
Today there are two ways to become a manager:
- Using `docker swarm join` and then `docker node promote`
- Using `docker swarm join --manager` directly

I think that's one too many ways and we should only support a single join scenario.

/cc @stevvooe @tonistiigi @icecrime

**Comment by tonistiigi** [comment]
I don't see them as conflicting and they are targeted to different users and use cases.

`join --manager` is safer, especially when used within scripts. If joining fails you will get an error right away and if it succeeds you can use the node for manager commands. This is especially useful when using secrets instead of manual acceptance and for bootstrapping clusters. For ops and people already familiar with the system.

`promote/demote` are nice for initial discovery of how managers work. For people who use small clusters that they manage manually. Demote is also only way to remove a node from raft members. When running `promote/demote` you never know if and when the action actually succeeded. Also it's a pain to use in scripts because you have to switch between machines.

I don't see this as any different from other cases where we have one command for create and another for update. You can set same things with `docker swarm init` that you can set with `docker swarm update`. `service create/update` etc.

**Comment by johnharris85** [comment]
Both of these ways make logical sense and apply to different use cases. I don't think having these two ways is terrible.

**Comment by aluzzardi** [comment]
This feature locks us down in a path.

I don't remember the specifics but I do remember a few times when this feature got in the way because of assumptions and we'll have to continue supporting it.

> If joining fails you will get an error right away

I see this as an issue rather than a feature :)

SwarmKit is entirely asynchronous. Joining a manager that is down (as a worker) is a completely valid use case. When a manager goes down, you don't want to block node provisioning.

**Comment by tonistiigi** [comment]
> SwarmKit is entirely asynchronous. Joining a manager that is down (as a worker) is a completely valid use case. When a manager goes down, you don't want to block node provisioning.

This is exactly how this works. We will tell user that join could not be completed because of some issue, so they can't expect to run manager commands yet. But we will also tell them that attempt to connect will continue in the background and how they can check if the node has become active.

**Comment by stevvooe** [comment]
In general, we want control of the topology to be done through the orchestration component. Having a node declare that it wants to be a manager upon joining is awkward from that perspective. With `join --manager`, topology is now controlled through "the script" rather than through the correct mechanism.

**Comment by aluzzardi** [comment]
I agree with @stevvooe.

The main thing for me is that we can always add `--manager` in a later release. However, we won't be able to remove it so it seems safer to get rid of it now.

**Comment by thaJeztah** [comment]
What I'm interested in, is a scenario where someone wants to automate setting up a cluster and spinning up workers, and managers; can we write out such a scenario and see if both can be achieved without the `--manager` flag

**Comment by dweomer** [comment]
I recently put together this Terraform script for standing up a 1.12 cluster:
```
variable "swarm_secret" {
  default = "@Sw4rm!"
}

variable "swarm_port" {
  default = 2377
}

resource "null_resource" "swarm_master" {
  count = 1

  depends_on = ["null_resource.requirements"]

  triggers {
    /* IF YOU CHANGE THE NAME OF THIS FILE PLEASE CHANGE THIS SELF REFERENCE */
    this_sha256 = "${sha256(file(format("%s/%s", path.module, "swarm.tf")))}"
  }

  connection {
    agent       = true
    user        = "${var.cluster_operator}"
    host        = "${element(split(",", module.provider_cluster.node_ipv4_addrs), 0)}"
    private_key = "${file(var.operator_ssh_key_file)}"
  }

  provisioner "remote-exec" {
    inline = [
      "docker swarm init --auto-accept manager --listen-addr 0.0.0.0:${var.swarm_port} --secret '${var.swarm_secret}'",
    ]
  }
}

resource "null_resource" "swarm_replica" {
  count = "4"

  depends_on = ["null_resource.swarm_master"]

  triggers {
    master_ip = "${element(split(",", module.provider_cluster.node_ipv4_addrs), 0)}"
  }

  connection {
    agent       = true
    user        = "${var.cluster_operator}"
    host        = "${element(split(",", module.provider_cluster.node_ipv4_addrs), count.index + 1)}"
    private_key = "${file(var.operator_ssh_key_file)}"
  }

  provisioner "remote-exec" {
    inline = [
      "docker swarm join --manager --secret ${var.swarm_secret} ${self.triggers.master_ip}:${var.swarm_port}",
    ]
  }
}
```

How does this work without `--manager` option during `join`? The only way that I see is making a second pass against the `swarm_manager` either after each individual replica `join` or after all such `joins` and ad-hod scripting it either way. While I agree with @aluzzardi in principle, in practice lacking this switch seems less than ideal. Maybe specify a manager secret and/or worker secret when initializing the first node?

**Comment by dweomer** [comment]
I meant to note: specifying `--auto-accept manager` without a corresponding `--secret` did not work when attempting to `join --manager`

**Comment by tiborvass** [comment]
This was fixed by #24823
